lib/server-app.js

```diff
@@ -161,7 +161,15 @@ proto.middleware = function(name, paths, handler) {
 
   this._skipLayerSorting = true;
   this.use(paths, handler);
-  this._router.stack[this._router.stack.length - 1].phase = fullPhaseName;
+
+  var layer = this._findLayerByHandler(handler);
+  if (layer) {
+    // Set the phase name for sorting
+    layer.phase = fullPhaseName;
+  } else {
+    debug('No matching layer is found for %s %s', fullPhaseName, handlerName);
+  }
+
   this._skipLayerSorting = false;
 
   this._sortLayersByPhase();
```

Comment:
Nitpick: when the layer was not found, there is no need to sort the layers. Feel free to ignore this comment.